src/utils/aadV2TokenProvider.ts

```diff
@@ -0,0 +1,278 @@
+  Scopes are always in $aadV2Token
+*/
+class AuthParameters {
+
+    private readonly aadV2TokenRegex: RegExp = new RegExp(`\\s*\\${Constants.AzureActiveDirectoryV2TokenVariableName}(\\s+(${Constants.AzureActiveDirectoryForceNewOption}))?(\\s+(appOnly))?(\\s+scopes:([\\w,.]+))?(\\s+tenantId:([^\\.]+\\.[^\\}\\s]+|[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}))?(\\s+clientId:([^\\.]+\\.[^\\}\\s]+|[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}))?\\s*`);
+    public tenantId: string;
+    public clientId: string;
+    public scopes: string[];
+    public forceNewToken: boolean;
+    public clientSecret?: string;
+    public appOnly: boolean;
+    public appUri?: string;
+
+    public constructor() {
+        this.clientId = AadV2TokenProviderClientId;
+        this.tenantId = "common";
+        this.forceNewToken = false;
+        this.appOnly = false;
+    }
+
+    async ReadEnvironmentVariable(variableName: string) : Promise<string | undefined> {
+        if (await EnvironmentVariableProvider.Instance.has(variableName)) {
+            const { value, error, warning } = await EnvironmentVariableProvider.Instance.get(variableName);
+            if (!warning && !error) {
+                return value as string;
+            } else {
+                return undefined;
+            }
+        }
+        return undefined;
+    }
+
+    GetCacheKey() : string {
+        return this.tenantId + "|" + this.clientId + "|" + this.appOnly as string;
+
+    }
+    async ParseName(name: string) {
+
+        // Update defaults based on environment
+        this.tenantId = (await this.ReadEnvironmentVariable("aadV2TenantId")) || this.tenantId;
+
+        let scopes = "openid,profile";
+        let explicitClientId: string|undefined = undefined;
+        // Parse variable parameters
+        const groups = this.aadV2TokenRegex.exec(name);
+        if (groups) {
+            this.forceNewToken = groups[2] === Constants.AzureActiveDirectoryForceNewOption;
+            this.appOnly = groups[4] === "appOnly";
+            scopes = groups[6] || scopes;
+            this.tenantId = groups[8] || this.tenantId;
+            explicitClientId = groups[10];
+        } else {
+            throw new Error("Failed to parse parameters: " + name);
+        }
+
+        this.scopes = scopes.split(",").map(s => s.trim());
+
+        if (this.appOnly) {
+            this.clientId = explicitClientId || (await this.ReadEnvironmentVariable("aadV2ClientId")) || this.clientId;
+            this.clientSecret = (await this.ReadEnvironmentVariable("aadV2ClientSecret"));
+            this.appUri = (await this.ReadEnvironmentVariable("aadV2AppUri"));
+            if (!(this.clientSecret && this.appUri)) {
+                throw new Error("For appOnly tokens, a environment variable aadV2ClientSecret and aadV2AppUri must be created.  aadV2ClientId and aadV2TenantId are optional environment variables.");
```

Comment:
It seems should be **environment variables** instead of **a environment variable**